### PR TITLE
EW-501 provided get all person 

### PR DESCRIPTION
**Pull request by Fshmit**
# Description
A get all persons endpoint was implemented. The request can be filtered with the following attributes: referrer, Familiennamen, Vorname, Sichtfeigabe.

## Links to Tickets or other PRs
https://ticketsystem.dbildungscloud.de/browse/EW-501

## Approval for review

- [ ] All points were discussed with the ticket creator, support-team or product owner. The code upholds all quality guidelines from the PR-template.

> Notice: Please remove the WIP label if the PR is ready to review, otherwise nobody will review it.

### CODE REVIEW

**Comment by janrenz** [comment]
I would add some pagination tests (or maybe I just overlooked them)

**Comment by Fshmit** [comment]
> I would add some pagination tests (or maybe I just overlooked them)
pagination will be implemented and tested in a separate ticket that is still in the backlog.

File: src/modules/person/persistence/person.repo.ts [hunk 1]
```diff
@@ -58,4 +57,29 @@ export class PersonRepo {
         await this.em.persistAndFlush(person);
         return this.mapper.map(person, PersonEntity, PersonDo);
     }
+
+    public async findAll(personDo: PersonDo<false>): Promise<Option<PersonDo<true>>[]> {
+        if (personDo.referrer !== undefined && personDo.isInformationBlocked !== undefined) {
+            const result: Option<PersonDo<true>>[] = await this.em.find(PersonEntity, {
```
Comment (pkleybolte) [inline src/modules/person/persistence/person.repo.ts:63]: `em.find` should return `PersonEntity[]`. No Option and no DO. You won't need the filter in line 80 anymore.

File: src/modules/person/persistence/person.repo.ts [hunk 2]
```diff
@@ -58,4 +57,29 @@ export class PersonRepo {
         await this.em.persistAndFlush(person);
         return this.mapper.map(person, PersonEntity, PersonDo);
     }
+
+    public async findAll(personDo: PersonDo<false>): Promise<Option<PersonDo<true>>[]> {
```
Comment (pkleybolte) [inline src/modules/person/persistence/person.repo.ts:61]: Option in the return Type does not make sense. The are no null or undefined entries in the array.

File: src/modules/person/api/person.uc.ts [hunk 3]
```diff
@@ -30,4 +31,18 @@ export class PersonUc {
...
         throw result.error;
     }
+
+    public async findAll(personDto: FindPersonDTO): Promise<PersonResponse[]> {
+        const personDo: PersonDo<false> = this.mapper.map(personDto, FindPersonDTO, PersonDo);
+        const result: Option<PersonDo<true>>[] = await this.personService.findAllPersons(personDo);
+        const persons: PersonResponse[] = [];
+        if (result) {
+            result.forEach((person: Option<PersonDo<true>>) => {
+                if (person !== null && person !== undefined) {
```
Comment (pkleybolte) [inline src/modules/person/api/person.uc.ts:41]: null or undefined check can be removed after return type of service changed.

File: src/modules/person/api/person.controller.spec.ts [hunk 4]
```diff
@@ -85,4 +96,58 @@ describe('PersonController', () => {
             expect(personUcMock.findPersonById).toHaveBeenCalledTimes(1);
         });
     });
+
+    describe('when getting all persons', () => {
```
Comment (pkleybolte) [inline src/modules/person/api/person.controller.spec.ts:100]: The description should just be 'findAll'.
We don't need a "when"-sentence for this.

File: src/modules/person/api/person.uc.spec.ts [hunk 5]
```diff
@@ -88,4 +91,27 @@ describe('PersonUc', () => {
...
     });
+
+    describe('findAll', () => {
+        const personDTO: FindPersonDTO = {
+            referrer: '',
+            familyName: '',
+            firstName: '',
+            visibility: false,
+        };
+        describe('when person exists', () => {
```
Comment (pkleybolte) [inline src/modules/person/api/person.uc.spec.ts:102]: when-describe not necessary

File: src/modules/person/domain/person.service.spec.ts [hunk 6]
```diff
@@ -118,4 +118,31 @@ describe('PersonService', () => {
             });
         });
     });
+
+    describe('findAllPersons', () => {
+        describe('when person exists', () => {
```
Comment (pkleybolte) [inline src/modules/person/domain/person.service.spec.ts:123]: No when-describes necessary

File: src/modules/person/domain/person.service.spec.ts [hunk 7]
```diff
@@ -118,4 +118,31 @@ describe('PersonService', () => {
...
+            });
+        });
+
+        describe('when no person were found', () => {
+            it('should return an empty list of persons ', async () => {
+                const person: PersonDo<false> = DoFactory.createPerson(false);
+                personRepoMock.findAll.mockResolvedValue([]);
+                mapperMock.map.mockReturnValue(person as unknown as Dictionary<unknown>);
+                const result: Option<PersonDo<true>>[] = await personService.findAllPersons(person);
+                expect(result).toBeInstanceOf(Array);
```
Comment (pkleybolte) [inline src/modules/person/domain/person.service.spec.ts:144]: missing expected length of 0

File: src/modules/person/persistence/person.repo.integration-spec.ts [hunk 8]
```diff
@@ -130,4 +130,37 @@ describe('PersonRepo', () => {
             });
         });
     });
+
+    describe('findAll', () => {
+        describe('when persons were found', () => {
```
Comment (pkleybolte) [inline src/modules/person/persistence/person.repo.integration-spec.ts:135]: no when-describes necessary

**Review by pkleybolte — CHANGES_REQUESTED** [review]

File: src/modules/person/persistence/person.repo.integration-spec.ts [hunk 9]
```diff
@@ -130,4 +130,33 @@ describe('PersonRepo', () => {
             });
         });
     });
+
+    describe('findAll', () => {
+        it('should delete person from database', async () => {
```
Comment (SimoneRadtke-Cap) [inline src/modules/person/persistence/person.repo.integration-spec.ts:135]: 'should find all persons from the database'

File: src/modules/person/persistence/person.repo.integration-spec.ts [hunk 10]
```diff
@@ -130,4 +130,33 @@ describe('PersonRepo', () => {
...
+            const props: Partial<PersonDo<false>> = {
+                referrer: 'referrer_value',
+                firstName: 'first name',
+                lastName: 'last name',
+                isInformationBlocked: false,
+            };
+            const personDo1: PersonDo<false> = DoFactory.createPerson(false, props);
+            const personDo2: PersonDo<false> = DoFactory.createPerson(false, props);
+            const personEntity1: PersonEntity = mapper.map(personDo1, PersonDo, PersonEntity);
```
Comment (SimoneRadtke-Cap) [inline src/modules/person/persistence/person.repo.integration-spec.ts:144]: You don't need personEntity1 and personEntity2. You can do
```
await em.persistAndFlush(mapper.map(personDo1, PersonDo, PersonEntity));
await em.persistAndFlush(mapper.map(personDo2, PersonDo, PersonEntity));
```

File: src/modules/person/persistence/person.repo.integration-spec.ts [hunk 11]
```diff
@@ -130,4 +130,33 @@ describe('PersonRepo', () => {
...
+            const personEntity2: PersonEntity = mapper.map(personDo2, PersonDo, PersonEntity);
+            await em.persistAndFlush(personEntity1);
+            await em.persistAndFlush(personEntity2);
+            const personDoFromQueryParam: PersonDo<false> = DoFactory.createPerson(false, props);
+            const result: PersonDo<true>[] = await sut.findAll(personDoFromQueryParam);
+            expect(result).not.toBeNull();
+            await expect(em.find(PersonEntity, {})).resolves.toHaveLength(2);
```
Comment (SimoneRadtke-Cap) [inline src/modules/person/persistence/person.repo.integration-spec.ts:151]: Why not 
```
expect(result).toHaveLength(2);
```
?

File: src/modules/person/persistence/person.repo.integration-spec.ts [hunk 12]
```diff
@@ -130,4 +130,33 @@ describe('PersonRepo', () => {
...
+        });
+
+        it('should return null', async () => {
```
Comment (SimoneRadtke-Cap) [inline src/modules/person/persistence/person.repo.integration-spec.ts:154]: Better 'should return an empty list'

File: src/modules/person/persistence/person.repo.integration-spec.ts [hunk 13]
```diff
@@ -130,4 +130,33 @@ describe('PersonRepo', () => {
...
+            const props: Partial<PersonDo<false>> = {};
+            const personDoFromQueryParam: PersonDo<false> = DoFactory.createPerson(false, props);
+            const result: PersonDo<true>[] = await sut.findAll(personDoFromQueryParam);
+            expect(result).not.toBeNull();
+            await expect(em.find(PersonEntity, {})).resolves.toHaveLength(0);
```
Comment (SimoneRadtke-Cap) [inline src/modules/person/persistence/person.repo.integration-spec.ts:159]: Maybe better:
```
expect(result).toEqual([]);
```

**Review by SimoneRadtke-Cap — CHANGES_REQUESTED** [review]

File: src/modules/person/api/person.controller.spec.ts [hunk 14]
```diff
@@ -85,4 +96,56 @@ describe('PersonController', () => {
             expect(personUcMock.findPersonById).toHaveBeenCalledTimes(1);
         });
     });
+
+    describe('findAll', () => {
```
[the remark this answers is not in the document]
Comment (Fshmit) [reply src/modules/person/api/person.controller.spec.ts:100]: it is the new test structure without using when statements.

File: src/modules/person/api/persons-query.param.spec.ts [hunk 15]
```diff
@@ -0,0 +1,22 @@
+import 'reflect-metadata';
+import { faker } from '@faker-js/faker';
+import { plainToInstance } from 'class-transformer';
+import { AllPersonsQueryParam } from './persons-query.param.js';
+describe('CreatePersonBodyParams', () => {
```
[the remark this answers is not in the document]
Comment (Fshmit) [reply src/modules/person/api/persons-query.param.spec.ts:5]: it is about the coverage :)

[the remark this answers is not in the document]
Comment (Fshmit) [reply src/modules/person/api/persons-query.param.spec.ts:5]: this file will be deleted. we don't need it even for the coverage

**Pull request closed without merging**